### PR TITLE
Update ldc to load Constant_Dynamic primitives as I_32 type

### PR DESCRIPTION
**Pull request by fengxue-IS**
ldc loads data into a UDATA slot on the stack and primitive values
only uses 32bit on a 64bit system, this change will ensure that
32 bit values will always be written to the lower 32 bit of the slot
(For BE and LE systems)

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

**Comment by fengxue-IS** [comment]
@DanHeidinga can you take a look?

File: runtime/vm/BytecodeInterpreter.hpp [hunk 1]
```diff
@@ -6124,24 +6124,33 @@ done:;
 			 */
 			switch (romCPEntry->cpType >> J9DescriptionReturnTypeShift) {
 			case J9DescriptionReturnTypeBoolean:
-				value = (j9object_t)(UDATA)J9VMJAVALANGBOOLEAN_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGBOOLEAN_VALUE(_currentThread, value);
```
Comment (DanHeidinga) [inline runtime/vm/BytecodeInterpreter.hpp]: I would prefer to use `(U_32)` to `(I_32)` - no accidental chance of sign extensions.  Though a boolean should be 0 or 1.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 2]
```diff
@@ -6124,24 +6124,33 @@ done:;
...
 				break;
 			case J9DescriptionReturnTypeByte:
-				value = (j9object_t)(UDATA)J9VMJAVALANGBYTE_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGBYTE_VALUE(_currentThread, value);
```
Comment (DanHeidinga) [inline runtime/vm/BytecodeInterpreter.hpp]: Should be `(I_32)(I_8) J9VMJAVALANGBYTE_VALUE `

File: runtime/vm/BytecodeInterpreter.hpp [hunk 3]
```diff
@@ -6124,24 +6124,33 @@ done:;
...
 				break;
 			case J9DescriptionReturnTypeChar:
-				value = (j9object_t)(UDATA)J9VMJAVALANGCHARACTER_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGCHARACTER_VALUE(_currentThread, value);
 				break;
 			case J9DescriptionReturnTypeShort:
-				value = (j9object_t)(UDATA)J9VMJAVALANGSHORT_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGSHORT_VALUE(_currentThread, value);
```
Comment (DanHeidinga) [inline runtime/vm/BytecodeInterpreter.hpp]: `(I_32)(I_16)`

File: runtime/vm/BytecodeInterpreter.hpp [hunk 4]
```diff
@@ -6124,24 +6124,33 @@ done:;
...
-				value = (j9object_t)(UDATA)J9VMJAVALANGBOOLEAN_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGBOOLEAN_VALUE(_currentThread, value);
 				break;
 			case J9DescriptionReturnTypeByte:
-				value = (j9object_t)(UDATA)J9VMJAVALANGBYTE_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGBYTE_VALUE(_currentThread, value);
 				break;
 			case J9DescriptionReturnTypeChar:
-				value = (j9object_t)(UDATA)J9VMJAVALANGCHARACTER_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGCHARACTER_VALUE(_currentThread, value);
```
Comment (DanHeidinga) [inline runtime/vm/BytecodeInterpreter.hpp]: `(I_32)(U_16)`

File: runtime/vm/BytecodeInterpreter.hpp [hunk 5]
```diff
@@ -6124,24 +6124,33 @@ done:;
...
 				break;
 			case J9DescriptionReturnTypeShort:
-				value = (j9object_t)(UDATA)J9VMJAVALANGSHORT_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGSHORT_VALUE(_currentThread, value);
 				break;
 			case J9DescriptionReturnTypeFloat:
-				value = (j9object_t)(UDATA)J9VMJAVALANGFLOAT_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGFLOAT_VALUE(_currentThread, value);
```
Comment (DanHeidinga) [inline runtime/vm/BytecodeInterpreter.hpp]: Use `(U_32)` instead of I_32 in both cases.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 6]
```diff
@@ -6124,24 +6124,33 @@ done:;
...
-				value = (j9object_t)(UDATA)J9VMJAVALANGINTEGER_VALUE(_currentThread, value);
+				*(I_32 *)_sp = (I_32)J9VMJAVALANGINTEGER_VALUE(_currentThread, value);
+				break;
+			case J9DescriptionReturnTypeDouble: /* fall through */
+			case J9DescriptionReturnTypeLong:
+				/* double and long value cannot be loaded by ldc as they require 2 slots */
+				Assert_VM_unreachable();
+				break;
+			default:
+				*_sp = (UDATA)value;
```
Comment (DanHeidinga) [inline runtime/vm/BytecodeInterpreter.hpp]: Why add the default when it duplicates the write below on line 6155?

Comment (fengxue-IS) [reply runtime/vm/BytecodeInterpreter.hpp]: this avoid having goto for each of the switch case

Comment (DanHeidinga) [reply runtime/vm/BytecodeInterpreter.hpp]: Fair enough.  The code will be ugly one or the other =)

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk8,11

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8,11

**Comment by gacholio** [comment]
Personally, I would have used a local for the 32-bit value, and not bothered adding the impossible cases to the case statement (that's real code being generated there unnecessarily).

**Comment by DanHeidinga** [comment]
@fengxue-IS Can you update as per GAC's suggestion?

**Comment by gacholio** [comment]
I suppose it's consistent with the rest of the code to assert in the default case (which is better than picking which cases are unimplmented). We can revisit this later on throughout the interpreter.

**Comment by fengxue-IS** [comment]
updated code, @gacholio can you take a look

**Comment by gacholio** [comment]
I'd still like to see:

- a single I_32 local and final assignment to sp
- default rather than the long cases for the assert

**Comment by gacholio** [comment]
Looks good - please squash (and I suggest adding [ci skip] into the squashed commit message).

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk8

**Pull request merged**